### PR TITLE
Contextual DotF predicate

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-1379

## Changes
- ~Add contextualizing support to mlang expression~
- ~Add ContextAwarePredicate model~
- Contextualize DotF predicate
- Fix for strategizer choice view choice selection in update view (included here as the change is just 2 lines)

### CODE REVIEW

**Review by nanoscott — APPROVED** [review]

File: src/foam/mlang/mlang.js [hunk 1]
```diff
@@ -3956,6 +3956,74 @@ foam.CLASS({
   ]
 });
 
+foam.CLASS({
```
Comment (jlhughes) [inline src/foam/mlang/mlang.js]: @kgrgreer going forward can we move new entries into their own file? Makes debugging in browser possible.

**Review by jlhughes — APPROVED** [review]

File: src/foam/core/ContextAware.js [hunk 2]
```diff
@@ -17,5 +17,21 @@ foam.INTERFACE({
...
       args: [ { name: 'x', type: 'Context' } ]
     }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.methods.push(`
+          static <T> T maybeContextualize(X x, T obj) {
```
Comment (kgrgreer) [inline src/foam/core/ContextAware.js:27]: FObject extends ContextAware, so when would obj ever not be an instanceof ContextAware?

File: src/foam/mlang/MLang.java [hunk 3]
```diff
@@ -37,9 +39,19 @@ public static Comparator DESC(Comparator c) {
   }
 
   public static Expr prepare(Object o) {
-    return o instanceof Expr ? (Expr) o :
+    var expr = o instanceof Expr ? (Expr) o :
```
Comment (kgrgreer) [inline src/foam/mlang/MLang.java]: I thought you had messed up JS and Java for  a minute when you used 'var'.

File: src/foam/mlang/mlang.js (hunk 1 above)
Comment (kgrgreer) [reply src/foam/mlang/mlang.js]: Sure.

File: src/foam/core/ContextAware.js (hunk 2 above)
Comment (lchanmann) [reply src/foam/core/ContextAware.js:27]: obj can be of any type, doesn't have to be FObject.

File: src/foam/mlang/MLang.java (hunk 3 above)
Comment (lchanmann) [reply src/foam/mlang/MLang.java]: Java 'var' correctly guesses that it is a `Expr` if not the compiler would yell on line 45.

**Comment by lchanmann** [comment]
Updated @kgrgreer @jlhughes please take another look. Thanks

**Pull request merged**